diff --git a/Docker/cu110/dockerfile b/Docker/cu110/dockerfile
index c1d91ad..4e359d6 100644
--- a/Docker/cu110/dockerfile
+++ b/Docker/cu110/dockerfile
@@ -15,7 +15,13 @@ RUN wget \
     && bash Miniconda3-latest-Linux-x86_64.sh -b \
     && rm -f Miniconda3-latest-Linux-x86_64.sh 
 RUN conda --version
-RUN conda install pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch -c conda-forge
+
+# @FIXME(bojian)  Change the PyTorch installation method. As the chosen one does
+#                 NOT come with CUDA support.
+# RUN conda install pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch -c conda-forge
+RUN pip install torch==1.8.0+cu111 torchvision==0.9.0+cu111 torchaudio==0.8.0 \
+        -f https://download.pytorch.org/whl/torch_stable.html
+
 RUN conda install -c dglteam dgl-cuda11.0
 RUN pip install requests
 
